src/libcore/ptr.rs

```diff
@@ -10,9 +10,49 @@
 
 // FIXME: talk about offset, copy_memory, copy_nonoverlapping_memory
 
-//! Raw, unsafe pointers, `*const T`, and `*mut T`.
+//! Manually manage memory through raw pointers.
 //!
 //! *[See also the pointer primitive types](../../std/primitive.pointer.html).*
+//!
+//! # Safety
+//!
+//! Many functions in this module take raw pointers as arguments and dereference
+//! them. For this to be safe, these pointers must be valid. However, because
+//! rust does not yet have a formal memory model, determining whether an
+//! arbitrary pointer is valid for a given operation can be tricky.
+//!
+//! There are two types of operations on memory, reads and writes. It is
+//! possible for a `*mut` to be valid for one operation and not the other. Since
+//! a `*const` can only be read and not written, it has no such ambiguity. For
+//! example, a `*mut` is not valid for writes if a a reference exists which
+//! [refers to the same memory][aliasing]. Therefore, each function in this
+//! module will document which operations must be valid on any `*mut` arguments.
+//!
+//! Additionally, some functions (e.g. [`copy`]) take a single pointer but
+//! operate on many values. In this case, the function will state the size of
+//! the operation for which the pointer must be valid. For example,
+//! `copy::<T>(&src, &mut dst, 3)` requires `dst` to be valid for writes of
+//! `size_of::<T>() * 3` bytes. When the documentation requires that a pointer
+//! be valid for an operation but omits the size of that operation, the size is
+//! implied to be `size_of::<T>()` bytes.
+//!
+//! For more information on the safety implications of dereferencing raw
+//! pointers, see the both the [book] and the section in the reference devoted
+//! to [undefined behavior][ub].
```

Comment:
I think that we can actually spell out the requirements *besides aliasing* here. As in, can't we start this section with the stuff we *know* -- that the range covered by `[pointer, pointer+size)` must be part of a single allocated memory block, and that the size is implicitly given by the type when not stated explicitly? Then we can say that there are additional requirements about aliasing, and that those rules are not settled yet, referencing the nomicon.

Reply:
I added some axioms regarding validity, casting a reference to a pointer results in a valid pointer, pointers to ZSTs are valid, that sort of thing, and linked to `offset` as a proxy for the single allocated memory block requirement. I also called out aliasing in that final paragraph.